### PR TITLE
Cache literal to number conversion.

### PR DESCRIPTION
**Pull request by chriseth**
Ticks one box in https://github.com/ethereum/solidity/issues/13201

Improves runtime on one example from 16.8 to 16 seconds.

### CODE REVIEW

File: libyul/Utilities.cpp [hunk 1]
```diff
@@ -91,9 +91,17 @@ u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
 {
 	yulAssert(_literal.kind == LiteralKind::Number, "Expected number literal!");
 
-	std::string const& literalString = _literal.value.str();
```
Comment (chriseth) [inline libyul/Utilities.cpp:94]: We could do this only for decimal numbers. Parsing a hex number is much faster than parsing a decimal number.

File: libyul/Utilities.cpp [hunk 2]
```diff
@@ -91,9 +91,17 @@ u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
...
-	yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-	return u256(literalString);
+	static map<YulString, u256> numberCache;
```
Comment (nikola-matic) [inline libyul/Utilities.cpp:94]: How large is this map expected to be? Would making it an `unordered_map` introduce any performance gains, or would it just introduce unnecessary overhead?

**Review by ekpyron — COMMENTED** [review]
If this actually makes a difference, wouldn't it be better to just store the literal value in the Yul AST node itself from the start?
Even if only as a ``mutable std::optional`` caching thing, if bookkeeping is too annoying otherwise...

File: libyul/Utilities.cpp [hunk 3]
```diff
@@ -91,9 +91,17 @@ u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
...
+	static YulStringRepository::ResetCallback callback{[&] { numberCache.clear(); }};
+
+	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
+	if (isNew)
+	{
+		std::string const& literalString = _literal.value.str();
```
Comment (nikola-matic) [inline libyul/Utilities.cpp:100]:
```suggestion
		string const& literalString = _literal.value.str();
```

**Comment by nikola-matic** [comment]
> Ticks one box in #13201
> 
> Improves runtime on one example from 16.8 to 16 seconds.

Just out of curiosity, how are you/we benchmarking this runtime? Are the results available to all via some CI job, or do you run it locally on your machine?

**Comment by chriseth** [comment]
@ekpyron this is more efficient than storing in the AST since it only needs to be done once per YulString.

File: libyul/Utilities.cpp (hunk 2 above)
Comment (ekpyron) [reply libyul/Utilities.cpp:94]: I'd be curious about this myself - but my guess would be that we don't have *that* many distinct literal values in the yul code for it to help. On the other hand, ``YulString``s being "prehashed" and all, always makes me wonder about it :-).

**Comment by ekpyron** [comment]
> @ekpyron this is more efficient than storing in the AST since it only needs to be done once per YulString.

For reference: result from chat: eventually it makes sense to change the main data of ``Literal``s to ``u256`` (merely keeping a string as optional formatting hint and resetting that on any value change), but we'll keep it at the cache here for now.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**